Sources/SwiftCuckoo/Session.swift

```diff
@@ -2,37 +2,74 @@ import Foundation
+        case idle                /// Indicates the session is not currently active.
+        case running             /// Indicates the session is currently in progress.
+        case completed           /// Indicates the session has completed.
+        case invalid(Invalid)    /// Indicates the session has an invalid state with a specific reason.
+    }
+    
     /// A unique identifier for the session instance.
     public var id: Identifier
     
     /// The start time of the session. This indicates when the session began.
-    public var startTime: Date
+    public var startTime: Date?
     
     /// The optional end time of the session. This indicates when the session
     /// was completed or could still be ongoing.
     public var endTime: Date?
 
-    /// Initializes a new session instance with a unique identifier, a start
-    /// time, and an optional end time.
+    /// Initializes a new session instance with a unique identifier and an optional start time.
     ///
     /// - Parameters:
-    ///   - id: A unique identifier for the session. This helps in managing
-    ///          and tracking individual session instances.
-    ///   - startTime: The starting time of the session, indicating when it begins.
-    ///   - endTime: The end time of the session, representing when the session is
-    ///              completed. This parameter is optional and defaults to nil for ongoing sessions.
+    ///   - id: A unique identifier for the session, aiding in tracking individual sessions.
+    ///   - startTime: The starting time of the session, indicating when it begins. Defaults to nil.
+    ///   - endTime: The end time of the session. This parameter is optional and defaults to nil.
     ///
-    /// - Note: Ensure that the `endTime` is provided if the session is completed.
-    package init(id: Identifier, startTime: Date, endTime: Date? = nil) {
+    /// - Note: Ensure that `endTime` is provided if the session is completed.
+    public init(id: Identifier) {
         self.id = id
-        self.startTime = startTime
-        self.endTime = endTime
+        self.startTime = nil
+        self.endTime = nil
+    }
+    
+    /// Starts the session at the specified date.
+    ///
+    /// - Parameter date: The date to start the session. Defaults to the current date.
+    /// - Throws: An error of type `Error` if:
+    ///   - The session is already running (cannot start twice).
+    public mutating func start(on date: Date = Date()) throws {
+        guard status() == .idle else {
+            throw Error.cannotStartTwice
+        }
+        
+        self.startTime = date
+    }
+    
+    /// Stops the session at the specified date.
+    ///
+    /// - Parameter date: The date to stop the session. Defaults to the current date.
+    /// - Throws: An error of type `Error` if:
+    ///   - The session has not been started (must start before stopping).
+    public mutating func stop(on date: Date = Date()) throws {
```

Comment:
I was thinking of going this route! I haven't played around with mutating functions much before this. However the reason we need the function to be mutating is because structs aren't mutable like classes due to being value types?

Reply:
Structs are value types instead of references. So you can change the original object, only copy (There is an internal mechanism to avoid unnecessary copies a.k.a copy-on-write). 
In that case, mutating methods will say to the compiler that the method will change internal state and the compiler will complain if it is a constant because it will need to make a copy and replace the current value.